File: 39-combination-sum/combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res=[]
        def backtrack(start,path,total):

            if total==target:
                res.append(list(path))
                return
            if total>target:
                return
            for i in range(start,len(candidates)):
                path.append(candidates[i])
                backtrack(i,path,total+candidates[i])
                path.pop()
        backtrack(0,[],0)
        return res
        
        

File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void solve(int sum,vector<int>& nums,int st,vector<int>& path){
        if(sum<0) return ;
        if(sum==0){
            ans.push_back(path);
            return;
        }
        
        for(int i=st;i<nums.size();i++){
            path.push_back(nums[i]);
            solve(sum-nums[i],nums,i,path);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n=candidates.size();
        vector<int> path;
        solve(target,candidates,0,path);
        return ans;
    }
};